### ISSUE TITLE
output Inf loss in training with metric of  gamma or gamma_deviance

### ISSUE THREAD

**Issue by chjq201410695** [Microsoft/LightGBM#2133]
## Environment info

Operating System: Win10

CPU/GPU model: CPU

C++/Python/R version: Phthon

LightGBM version or commit hash: 2.2.2

## model info
params = { 'task':'train',
             'boosting_type': 'dart', 
             'objective': 'gamma',
             'learning_rate': 0.001, 
             'num_leaves':80,
             'min_sum_hessian_in_leaf': 0.001,
             'max_depth': 7, 
             'min_data_in_leaf':15,
             'tree_learner':'data',
             'num_threads': -1,
             'bagging_fraction': 0.8,
             'feature_fraction': 0.9,
             'verbose': 1,
             'max_bin' : 255,
             'min_data_in_bin': 15,
             'metric': {'gamma','gamma_deviance','l1'},
             'lambda_l1':0.2,
             'lambda_l2':0.15,
             'drop_rate':0.4,
            'num_iterations' 20000,
            'early_stopping_round'500,
            }
model_AX_reg = lgb.train(params,
                  train_set = lgb_train,
                  valid_sets = [lgb_train,lgb_eval],
                  valid_names = ["train","test"],
                  verbose_eval = 100,
                  categorical_feature= ['UN_x','UN_y','TS_x','TS_y','ICS_x','ICS_y','MS_x','MS_y','od_metro'],
                  evals_result=evals_result  #这条很重要，记录训练过程
                  )

## Error message
Training until validation scores don't improve for 500 rounds.
[100]	train's gamma-deviance: inf	train's l1: 0.460296	train's gamma: nan	test's gamma-deviance: inf	test's l1: 0.456405	test's gamma: nan
[200]	train's gamma-deviance: inf	train's l1: 0.468636	train's gamma: nan	test's gamma-deviance: inf	test's l1: 0.46448	test's gamma: nan
[300]	train's gamma-deviance: inf	train's l1: 0.472344	train's gamma: nan	test's gamma-deviance: inf	test's l1: 0.468244	test's gamma: nan

you can see that model is training but gamma and gamma-deviance are inf here.

Is there any idea?

update
-------------------------------------------
It seems like there are log(0)s in y.    In doc, the illustration of Gamma regression is Gamma regression with log-link. I wonder the log-link is np.log() or np.log1p()?

**Comment by chjq201410695** [comment]
I find the issue from log(0), given ""

**Comment by chjq201410695** [comment]
the
